### PR TITLE
Next GraphQL client

### PR DESCRIPTION
**Pull request by plmercereau**

### CODE REVIEW

File: packages/graphql-js/tests/unit/documents.test.ts [hunk 1]
```diff
@@ -0,0 +1,232 @@
+import { print } from 'graphql/language/printer'
+import { describe, expect, it } from 'vitest'
+import { toGraphQLDocument } from '../../src/graphql-object/query-builder'
+import schema from '../schemas/hasura'
+
+describe('main', () => {
```
Comment (szilarddoro) [inline packages/graphql-js/tests/unit/documents.test.ts:6]: could you please add some tests to cover cases where multiple `where` arguments are used?

File: packages/nextjs/nextjs.typedoc.json [hunk 2]
```diff
@@ -26,12 +26,12 @@
       "@nhost/hasura-storage-js": [
         "../hasura-storage-js/src/index.ts"
       ],
+      "@nhost/nhost-js": [
```
Comment (szilarddoro) [inline packages/nextjs/nextjs.typedoc.json:29]: why was this change necessary?

File: packages/react/react.typedoc.json [hunk 3]
```diff
@@ -26,11 +26,14 @@
...
         "../hasura-storage-js/src/index.ts"
       ],
+      "@nhost/graphql-js": [
+        "../graphql-js/src/index.ts"
+      ],
       "@nhost/nextjs": [
         "../nextjs/src/index.ts"
       ],
       "@nhost/nhost-js": [
-        "../nhost-js/src/index.ts"
```
Comment (szilarddoro) [inline packages/react/react.typedoc.json:33]: why is this change necessary?

File: packages/vue/vue.typedoc.json [hunk 4]
```diff
@@ -26,11 +26,14 @@
...
         "../hasura-storage-js/src/index.ts"
       ],
+      "@nhost/graphql-js": [
+        "../graphql-js/src/index.ts"
+      ],
       "@nhost/nextjs": [
         "../nextjs/src/index.ts"
       ],
       "@nhost/nhost-js": [
-        "../nhost-js/src/index.ts"
```
Comment (szilarddoro) [inline packages/vue/vue.typedoc.json:33]: why is this change necessary?

File: packages/hasura-storage-js/src/hasura-storage-api.ts [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import fetch from 'cross-fetch'
 
+import { fetchUpload } from './utils'
```
Comment (elitan) [inline packages/hasura-storage-js/src/hasura-storage-api.ts:3]: Why are there changes to hasura-storage-api in this PR?

File: packages/hasura-storage-js/src/utils/index.ts [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import { StorageImageTransformationParams } from './types'
 
 export * from './types'
+export * from './upload'
```
Comment (elitan) [inline packages/hasura-storage-js/src/utils/index.ts:4]: Why are there changes to hasura-storage-api in this PR?

**Review by elitan — CHANGES_REQUESTED** [review]

File: packages/graphql-js/functions/index.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
+import SchemaBuilder from '@pothos/core'
```
[the remark this answers is not in the document]
Comment (elitan) [reply packages/graphql-js/functions/index.ts:1]: This function is exposed as a remote schema in Hausra and is used for e2e tests in `tests/e2e/graphql.test.ts`.

File: packages/hasura-storage-js/src/utils/index.ts (hunk 6 above)
Comment (plmercereau) [reply packages/hasura-storage-js/src/utils/index.ts]: `toIso88591` was not exposed for tests. I reverted.

File: packages/nextjs/nextjs.typedoc.json (hunk 2 above)
Comment (plmercereau) [reply packages/nextjs/nextjs.typedoc.json:29]: for the same reason we discussed about: typedoc fails with a heap out of memory error. When setting the path to the transpilted dts, typedoc succeeds

File: packages/react/react.typedoc.json (hunk 3 above)
Comment (plmercereau) [reply packages/react/react.typedoc.json:33]: same as the above

File: packages/vue/vue.typedoc.json (hunk 4 above)
Comment (plmercereau) [reply packages/vue/vue.typedoc.json:33]: same as the above

File: packages/hasura-storage-js/src/hasura-storage-api.ts (hunk 5 above)
Comment (plmercereau) [reply packages/hasura-storage-js/src/hasura-storage-api.ts]: `toIso88591` was not exposed for tests. I reverted.

File: packages/graphql-js/tests/unit/documents.test.ts (hunk 1 above)
Comment (plmercereau) [reply packages/graphql-js/tests/unit/documents.test.ts:6]: done

File: packages/nextjs/nextjs.typedoc.json (hunk 2 above)
Comment (szilarddoro) [reply packages/nextjs/nextjs.typedoc.json:29]: Would this change solve the issue for `nhost-js`'s docgen as well? Now all of the references to docgen were removed from `nhost-js`, but I'd assume that pointing everything to transpiled code would solve the issue there too. Have you tried doing that?

**Review by elitan — CHANGES_REQUESTED** [review]

File: packages/graphql-js/tests/e2e/hasura.test.ts [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        "updatedAt",
+        "userId",
+      ]
+    `)
+  })
+
+  it('should ignore an invalid property', async () => {
+    const todos = await client.query.todos({
+      select: { id: true, unexistingProperty: true }
+    })
```
[the remark this answers is not in the document]
Comment (plmercereau) [reply packages/graphql-js/tests/e2e/hasura.test.ts:102]: `unexistingProperty` is ignored

[the remark this answers is not in the document]
Comment (elitan) [reply packages/graphql-js/tests/e2e/hasura.test.ts:102]: Hench my question.

[the remark this answers is not in the document]
Comment (plmercereau) [reply packages/graphql-js/tests/e2e/hasura.test.ts:102]: For the moment, the query builder sees `unexistingProperty` does not exist and therefore generates `todos { id }`.
We can:
1. generate `todos { id }`
2. generate `todos { id, unexistingProperty}` - the query is then sent to the server, which returns the error
3. return an error.

Do you want to switch from 1 to 3?

[the remark this answers is not in the document]
Comment (szilarddoro) [reply packages/graphql-js/tests/e2e/hasura.test.ts:102]: I think this should be a type error.

File: packages/nextjs/nextjs.typedoc.json (hunk 2 above)
Comment (plmercereau) [reply packages/nextjs/nextjs.typedoc.json:29]: It passes the build, but I haven't checked all the links between markdown files. That's why I avoided setting it that way everywhere. For instance, if an interface is defined in `@nhost/hasura-auth-js` but the `@nhost/hasura-auth-js` path resolves to the transpiled dts, then the markdown file that describes this interface won't be rendered nicely.

File: packages/graphql-js/tests/e2e/hasura.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (plmercereau) [reply packages/graphql-js/tests/e2e/hasura.test.ts:102]: > I think this should be a type error.

We talked about the fact I could not overcome this two weeks ago. The result is, however, correctly typed. Related: https://github.com/microsoft/TypeScript/issues/26242

[the remark this answers is not in the document]
Comment (elitan) [reply packages/graphql-js/tests/e2e/hasura.test.ts:102]: I think option #3 is better than option #1.

[the remark this answers is not in the document]
Comment (plmercereau) [reply packages/graphql-js/tests/e2e/hasura.test.ts]: > I think option 3 is better than option 1.

done

**Review by szilarddoro — APPROVED** [review]

File: packages/graphql-js/tests/codegen.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+generates:
```
[the remark this answers is not in the document]
Comment (plmercereau) [reply packages/graphql-js/tests/codegen.yaml:1]: both works. For the tests I preferred using the `yaml` syntax so there is no additional effort to put in making sure tsconfig is correctly set.

**Review by elitan — APPROVED** [review]
